src/plugins/visualizations/public/vis_types/base_vis_type.ts

```diff
@@ -42,9 +41,20 @@ export interface BaseVisTypeOptions {
   setup?: unknown;
   useCustomNoDataScreen?: boolean;
   inspectorAdapters?: Adapters | (() => Adapters);
-  toExpressionAst?: VisToExpressionAst;
 }
 
+interface ExpressionBaseVisTypeOptions extends CommonBaseVisTypeOptions {
+  toExpressionAst: VisToExpressionAst;
+  visualization?: undefined;
+}
+
+interface VisualizationBaseVisTypeOptions extends CommonBaseVisTypeOptions {
+  toExpressionAst?: undefined;
+  visualization: VisualizationControllerConstructor | undefined;
+}
+
+export type BaseVisTypeOptions = ExpressionBaseVisTypeOptions | VisualizationBaseVisTypeOptions;
```

Comment:
:information_source: Properly splitted this up into two types. It either needs to have a `visualization` or a `toExpressionAst`, but not both. Technically this COULD still be used that way (since the `toExpressionAst` could return an expression containing the `visualization` function, that would then go and look for `visualization`), but we're not using it this way, since `toExpressionAst` is used for the functions using their own renderer, so I rather wanted this state to reflect the refactoring we want to end up with, not what technically theoretically would be possible.